File: src/go/problems/pair_sum_test.go
// https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list

package problems

import (
	"testing"

	"github.com/dobermanch/leetcode/core"
)

type PairSum struct{}

func TestPairSum(t *testing.T) {
	gen := core.TestSuite[PairSum]{}
	gen.Add(func(tc *core.TestCase) {
		l1 := ListNode{
			Val: 5,
			Next: &ListNode{
				Val: 4,
				Next: &ListNode{
					Val: 2,
					Next: &ListNode{
						Val: 1,
					},
				},
			},
		}

		tc.Param(l1).Result(6)
	}).Run(t)
}

func (PairSum) Solution(head *ListNode) int {
	slow := head
	fast := head

	for fast != nil && fast.Next != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}

	var tail *ListNode
	for slow != nil {
		next := slow.Next
		slow.Next = tail
		tail = slow
		slow = next
	}

	max := 0
	for tail != nil {
		sum := tail.Val + head.Val
		if sum > max {
			max = sum
		}

		head = head.Next
		tail = tail.Next
	}

	return max
}

File: src/python/problems/pair_sum_test.py
# https://leetcode.com/problems/maximum-twin-sum-of-a-linked-list

from typing import Optional
from core.problem_base import *
from models.list_node import *

class DeleteMiddle(ProblemBase):
    def Solution(self, head: Optional[ListNode]) -> int:
        if not head.next:
            return None

        slow = head
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        tail = None
        while slow:
            next = slow.next
            slow.next = tail
            tail = slow
            slow = next
        
        result = 0
        while tail:
            result = max(result, tail.val + head.val)
            head = head.next
            tail = tail.next
        
        return result


if __name__ == '__main__':
    TestGen(DeleteMiddle) \
        .Add(lambda tc: tc.Param(ListNode(1, ListNode(4, ListNode(2, ListNode(1))))).Result(6)) \
        .Run()